torchhd/datasets/dataset.py

```diff
@@ -0,0 +1,331 @@
+import os
+import os.path
+from typing import Callable, Optional, Tuple, List
+import torch
+from torch import Tensor
+from torch.utils import data
+import pandas as pd
+import tarfile
+import numpy as np
+import torchhd
+
+
+from .utils import download_file_from_google_drive
+
+
+# Function that iterates via a given collection of datasets
+def UCIDatasetCollection(dataset_collection, root, download):
+    # For all datasets in the collection
+    for i in range(len(dataset_collection)):
+        # Fetch the current dataset
+        dataset = getattr(torchhd.datasets, dataset_collection[i])
+
+        train_ds = []
+        test_ds = []
+        # If no separate test dataset available - do 4-fold cross-validation
+        if hasattr(dataset, "num_folds"):
+            for fold_id in range(dataset.num_folds):
+                # Set test and train datasets for the current fold
+                train_ds.append(
+                    dataset(root, train=True, download=download, fold=fold_id)
+                )
+                test_ds.append(dataset(root, train=False, download=False, fold=fold_id))
+
+        # Case of avaiable test set
+        else:
+            # Set test and train datasets
+            train_ds.append(dataset(root, train=True, download=download))
+            test_ds.append(dataset(root, train=False, download=False))
+
+        yield train_ds, test_ds
```

Comment:
It could be that I do not fully understanding your original idea @mikeheddes. But when testing the original version of the function, for the datasets that need to be cross-validated it would always return only the last (4th) fold so, therefore, I was looking for some work around to address this issue and came up with the idea or returning the list of datasets since that way one is agnostic about whether there single train/test split or whether there are 4 such splits.

Reply:
Ah my bad I had a bug in my example, the yield of datasets should have been within the for loop over folds. Sorry!

Reply:
My idea was that from the user's perspective you would just see one dataset at a time and you don't really care so much which fold it is.